Automattic/wp-calypso
# Consume monorepo packages without building them

This PR makes it possible to consume monorepo packages (`packages/*`) without building them. To build Calypso, i.e., its webpack bundle, the `dist/esm` and `dist/cjs` folders don't need to be created and populated with transpiled files. That's a task that's needed only when preparing a package to be published in the NPM registry.

Instead, the Calypso webpack build can now directly consume the sources in `src/`, and transpile any Babel-enhanced JavaScript or TypeScript with its webpack loaders, just like any other non-package file in `client/`.

I achieved this by adding `calypso:main` field to all packages, and pointing in to the `./src/index` source file. Then I configured the webpack resolver to look at `calypso:main` first, using the `resolve.mainFields` option.

To make tests work, the Jest resolver needs to be aware of `calypso:main`, too. That's harder, and I had to write a custom resolver (`packages/calypso-build/jest/calypso-resolver`) by copying and modifying the default resolver from Jest sources.

Building Calypso, both in development (with the dev middleware that watches the sources for changes) and production, now doesn't require the `build-packages` or `build-packages:watch` steps. Simply running webpack is enough.

The only package that still needs the build step is `calypso-color-schemes`. It does SCSS-to-CSS-and-JS compilation, and also performs some magic with copying `color-schemes` assets (I remember that @ockham had to add that at some point to solve some problem that didn't have any other solution)

**Risks:**
There are several independent implementations of the module resolution algorithm in various tools:
- webpack
- TypeScript
- Jest
- Node.js

And not all of them can be easily configured to look at the `calypso:main` fields. We might end up with some part of our build pipeline being unable to load the monorepo packages correctly.

**How to test:**
Run `yarn start` and see if it builds Calypso correctly and if the build is faster now.
